binderhub/builder.py

```diff
@@ -408,31 +408,69 @@ async def get(self, provider_prefix, _unescaped_spec):
             else:
                 image_found = True
 
+        require_build_only = self.settings.get("require_build_only", False)
+        build_only = False
+        if not require_build_only:
+            build_only_query_parameter = self.get_query_argument(
+                name="build_only", default=""
+            )
+            if build_only_query_parameter.lower() == "true":
+                raise HTTPError(
+                    log_message="Building but not launching is not permitted!"
+                )
+        else:
+            # Not setting a default will make the function raise an error
+            # if the `build_only` query parameter is missing from the request
+            build_only_query_parameter = self.get_query_argument(name="build_only")
+            if build_only_query_parameter.lower() != "true":
+                raise HTTPError(
+                    log_message="The `build_only=true` query parameter is required!"
+                )
+            # If we're here, it means a build only deployment is required
+            build_only = True
+
+        if build_only:
+            await self.emit(
+                {
+                    "phase": "info",
+                    "imageName": image_name,
+                    "message": "Build only was enabled. Image will not be launched after build.\n",
+                }
+            )
         if image_found:
             await self.emit(
                 {
                     "phase": "built",
                     "imageName": image_name,
-                    "message": "Found built image, launching...\n",
+                    "message": "Found built image\n",
                 }
             )
             with LAUNCHES_INPROGRESS.track_inprogress():
-                await self.launch(provider)
-            self.event_log.emit(
-                "binderhub.jupyter.org/launch",
-                5,
-                {
-                    "provider": provider.name,
-                    "spec": spec,
-                    "ref": ref,
-                    "status": "success",
-                    "build_token": self._have_build_token,
-                    "origin": self.settings["normalized_origin"]
-                    if self.settings["normalized_origin"]
-                    else self.request.host,
-                },
-            )
-            return
+                if build_only:
+                    await self.emit(
+                        {
+                            "phase": "ready",
+                            "imageName": image_name,
+                            "message": "Image won't be launched\n",
+                        }
+                    )
```

Comment:
https://github.com/jupyterhub/binderhub/blob/554549c5ff83d87c57e7bb88b05c0120f1b17afe/binderhub/builder.py#L55-L58

It seems `LAUNCHES_INPROGRESS` is about a prometheus metric tracking launch attempts. I figure we should not run code associated with build_only under it as it may briefly then increment a counter (to later quickly decrement it).